src/Evolution/DiscontinuousGalerkin/Actions/InternalMortarDataImpl.hpp

```diff
@@ -230,53 +229,110 @@ void internal_mortar_data_impl(
+              std::optional{std::pair{face_mesh, DataVector{}}};
+        }
+
+        // Always set the time step ID
+        mortar_data_ptr->at(mortar_id).time_step_id() = temporal_id;
+
+        auto& local_mortar_data = local_mortar_data_opt->second;
+
+        // Do a destructive resize to account for potential p-refinement
+        local_mortar_data.destructive_resize(total_face_size);
+
+        ASSERT(local_mortar_data.size() == total_face_size,
+               "AhHHhhhhHHHHhHhhhHHh");
+        packaged_data.set_data_ref(local_mortar_data.data(),
+                                   local_mortar_data.size());
+      }
+    } else {
+      // In this case, we have multiple neighbors in this direction so all will
+      // need to project their data which means we use the
+      // mortar_projection_buffer to calculate the dg_package_data
+      packaged_data.set_data_ref(mortar_projection_buffer->data(),
+                                 total_face_size);
+    }
+
+    detail::dg_package_data<System>(
         make_not_null(&packaged_data), boundary_correction, fields_on_face,
         get<evolution::dg::Tags::NormalCovector<Dim>>(
             *normal_covector_and_magnitude_ptr->at(direction)),
         face_mesh_velocity, dg_package_data_projected_tags{},
         package_data_volume_args...);
-    (void)max_abs_char_speed_on_face;
 
     // Perform step 3
+    // This will only do something if
+    //  a) we have multiple neighbors in this direction
+    //  b) the one (and only) neighbor in this direction needed projection
     for (const auto& neighbor : neighbors_in_direction) {
       const auto mortar_id = std::make_pair(direction, neighbor);
       const auto& mortar_mesh = mortar_meshes.at(mortar_id);
       const auto& mortar_size = mortar_sizes.at(mortar_id);
 
-      // Project the data from the face to the mortar if necessary.
-      // We can't move the data or modify it in-place when projecting, because
-      // in that case the face may touch two mortars so we need to keep the
-      // data around.
-      DataVector boundary_data_on_mortar{};
       if (Spectral::needs_projection(face_mesh, mortar_mesh, mortar_size)) {
-        boundary_data_on_mortar = DataVector{
+        auto& local_mortar_data_opt =
+            mortar_data_ptr->at(mortar_id).local_mortar_data();
+
+        // If this isn't the first time, set the face mesh
+        if (LIKELY(local_mortar_data_opt.has_value())) {
+          local_mortar_data_opt->first = face_mesh;
+        } else {
+          // If we don't do this, then the DataVector will be non-owning which
+          // we don't want
+          local_mortar_data_opt =
+              std::optional{std::pair{face_mesh, DataVector{}}};
+        }
+
+        // Set the time id since above we only set it for cases that didn't need
+        // projection
+        mortar_data_ptr->at(mortar_id).time_step_id() = temporal_id;
+
+        auto& local_mortar_data = local_mortar_data_opt->second;
```

Comment:
Would you mind doing `DataVector&` here? I love `auto` as much as anyone else, but in some cases I find it does hinder readability